### PR TITLE
Let MySQLConnector accept all UTF charsets and recommend utf8mb4

### LINKED ISSUE #5377: mysql connector should (probably) allow utf8mb4 as character set
I tried to set up druid with mysql as metadata storage. It has utf8mb4 as default character set, which currently seems to be the [recommended best practice](https://stackoverflow.com/questions/766809/whats-the-difference-between-utf8-general-ci-and-utf8-unicode-ci). Druid rejected it with "Database default character set is not UTF-8." [here](https://github.com/druid-io/druid/blob/0105cdbc19828009d21de57a30ba55794a518d30/extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java#L167).
I guess, without further testing, that this is too restrictive and druid should allow utf8mb4.

### PR DESCRIPTION
**Pull request by hoesler**
Allow and prefer utf8mb4 character set in mysql metadata storage.

Upgrade Instructions:
Use ALTER DATABASE and ALTER TABLE as described in https://dev.mysql.com/doc/refman/5.7/en/charset-unicode-conversion.html

Closes #5377

### CODE REVIEW

File: extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java [hunk 1]
```diff
@@ -163,17 +162,19 @@ protected int getStreamingFetchSize()
...
+        .createQuery("SELECT @@character_set_database")
+        .map(StringMapper.FIRST)
         .first();
 
-    if (!isUtf8) {
+    if (!databaseCharset.matches("utf.*")) {
       throw new ISE(
-          "Database default character set is not UTF-8." + System.lineSeparator()
-          + "  Druid requires its MySQL database to be created using UTF-8 as default character set."
+          "Druid requires its MySQL database to be created with an UTF charset, found `%1$s`.%n" +
```
Comment (drcrallen) [inline extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java]: Can you remove newlines ? It confuses lots of logging systems.

File: extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java [hunk 2]
```diff
@@ -163,17 +162,19 @@ protected int getStreamingFetchSize()
...
-    boolean isUtf8 = handle
-        .createQuery("SELECT @@character_set_database = 'utf8'")
-        .map(BooleanMapper.FIRST)
+    String databaseCharset = handle
+        .createQuery("SELECT @@character_set_database")
+        .map(StringMapper.FIRST)
         .first();
 
-    if (!isUtf8) {
+    if (!databaseCharset.matches("utf.*")) {
```
Comment (drcrallen) [inline extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java]: Is the result ever cased differently? Do we need a to lowercase?

Comment (drcrallen) [reply extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java]: Don't this also match a bunch of other things like utf16?

Comment (hoesler) [reply extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java]: Don't you think lowercase is okay, if it has been until now? But yes, it should be `uft8.*`. I'll fix it.

File: extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java (hunk 1 above)
Comment (hoesler) [reply extensions-core/mysql-metadata-storage/src/main/java/io/druid/metadata/storage/mysql/MySQLConnector.java]: Sure.

**Comment by drcrallen** [comment]
Cool, https://github.com/druid-io/druid/blob/druid-0.11.0/docs/content/development/extensions-core/mysql.md needs updated, and can you add "upgrade instructions" to the PR master comment for those who used the "old" instructions for utf8 mysql databases?

**Comment by hoesler** [comment]
I already updated the mysql.md. Is anything missing there?

For the upgrade instructions, I added a link to the mysql docs, as it is not generic. See here: https://dba.stackexchange.com/questions/8239/how-to-easily-convert-utf8-tables-to-utf8mb4-in-mysql-5-5

**Comment by hoesler** [comment]
Alternatively, I could assemble all required statements to migrate in the mysql.md.

**Comment by drcrallen** [comment]
Cool! 

@hoesler would you mind filling out http://druid.io/community/cla.html please?

**Comment by hoesler** [comment]
Sorry for the delay. Signed it!

**Comment by gianm** [comment]
@drcrallen I'm assuming you are good due to the positivity of your last comment (https://github.com/druid-io/druid/pull/5411#issuecomment-368383317). Let us know if otherwise.

**Pull request merged**